packages/suite/src/validate-test.ts

```diff
@@ -19,6 +19,15 @@ export class TestValidationError extends Error {
   }
 }
 
+const every = Array.prototype.every;
+const some = Array.prototype.some;
+
+function validateTestKeys (test: Test, keys: (keyof Test)[], validationFn: typeof some | typeof every): boolean {
+  // the disable comment below is because eslint is not recognising k as used in !!test?.[k].
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  return validationFn.call(keys, (k: keyof Test) => !!test?.[k]);
+}
+
```

Comment:
IMO this is a bit too clever. I've stared at this function for ten minutes and I'm still not sure what it does 😂

An alternative suggestion: how about changing the signature of `validateTest` to take an `unknown` instead, this actually meshes with how it is used, since we want to validate completely arbitrary input, we could then do something like:


``` typescript
// runtime validation and casting of the `Test` type
function validateTestType(test: any): Test {
  if(!test) {
     throw new TestValidationError("Invalid Test: the test is empty, did you export anything?")
  }
  if(!test.description) {
     throw new TestValidationError("Invalid Test: the test must have a description")
  }
  // ...
  return test as Test;
}

function validateTest(input: unknown) {
  let test = validateTestType(input);
  // ...
}
```

Reply:
> IMO this is a bit too clever. I've stared at this function for ten minutes and I'm still not sure what it does 😂
> 

haha, I'm not sure `clever` is the right word here if nobody gets what it does.   I think you have a point

You could have `Partial<Test>` rather than unknown:

```ts
unction validateTestType(test?: Partial<Test>): Test {
```

Reply:
I think that would be correct at runtime, but might be slightly less accurate from a typing perspective, since we could in theory export something like a number from a test file, and then we would end up calling this with a number, which I think should work at runtime, but would technically be not allowed by the TS typesystem.

Reply:
> since we could in theory export something like a number from a test file
Exporting a number would be incorrect though?

I would say we are stating here what the expected type is and not what could be passed.

Typescript has no runtime semantics

Reply:
Yes, precisely. Since the input to this function is provided by the end user via whatever they export from their testfiles, it could be literally anything, even something silly like a number. The `validateTestType` function I sketched out above is a runtime implementation of type checking. It will check all properties of the type and only if they are satisfied will it return a properly typecast value.

Reply:
> Yes, precisely. Since the input to this function is provided by the end user via whatever they export from their testfiles, it could be literally anything, even something silly like a number. The `validateTestType` function I sketched out above is a runtime implementation of type checking. It will check all properties of the type and only if they are satisfied will it return a properly typecast value.

Makes sense then what you are suggesting might be better as a type guard:

```ts
function validateTestType(test?: Partial<Test>): test is Test {
```

Reply:
Yes, that totally works too, might be tricky though since it needs to throw in case of error, not return false.